File: python/tests/test_branches.py
import pandas as pd

from rts_gmlc_case import data
from rts_gmlc_case.branches import add_branches
from rts_gmlc_case.document import CaseDocument
from rts_gmlc_case.topology import add_topology


def test_branch_partition_and_arcs():
    doc = CaseDocument()
    src = data.rts_source_dir() / "SourceData"
    idx = add_topology(doc, src)
    add_branches(doc, src, idx)

    br = pd.read_csv(src / "branch.csv")
    dc = pd.read_csv(src / "dc_branch.csv")
    bus = pd.read_csv(src / "bus.csv")
    base_kv = dict(zip(bus["Bus ID"], bus["BaseKV"]))

    # Two independent classification rules must agree exactly (R21).
    tr_ratio_rule = ~br["Tr Ratio"].isin([0.0, 1.0])
    voltage_rule = br.apply(
        lambda row: base_kv[row["From Bus"]] != base_kv[row["To Bus"]], axis=1
    )
    assert (tr_ratio_rule == voltage_rule).all()
    n_xf = int(tr_ratio_rule.sum())

    assert len(doc.document.components["Line"]) == len(br) - n_xf
    assert len(doc.document.components["TwoWindingTransformer"]) == n_xf
    assert len(doc.document.components["TransformerCircuit"]) == n_xf

    # Pinned expected totals (CAMPAIGN-FACTS.md "Expected component counts", R21).
    assert len(doc.document.components["Line"]) == 105
    assert len(doc.document.components["TransformerCircuit"]) == 15
    assert len(doc.document.components["TwoWindingTransformer"]) == 15
    assert len(doc.document.components["TwoTerminalGenericHVDCLine"]) == 1

    pairs = [(a["from_id"], a["to_id"]) for a in doc.document.components["Arc"]]
    assert len(set(pairs)) == len(pairs)  # deduplicated

    ac_pairs = {
        (idx.bus_id_by_number[int(row["From Bus"])], idx.bus_id_by_number[int(row["To Bus"])])
        for row in br.to_dict("records")
    }
    dc_pairs = {
        (idx.bus_id_by_number[int(row["From Bus"])], idx.bus_id_by_number[int(row["To Bus"])])
        for row in dc.to_dict("records")
    }
    assert len(ac_pairs) == 108
    assert len(dc_pairs) == 1
    assert len(ac_pairs | dc_pairs) == len(doc.document.components["Arc"])
    assert len(doc.document.components["Arc"]) == 109

    # No transformer circuit's name field (it has none) and no line is missing an arc.
    for line in doc.document.components["Line"]:
        assert line["arc"] in {a["id"] for a in doc.document.components["Arc"]}
    for circuit in doc.document.components["TransformerCircuit"]:
        assert "name" not in circuit
    xf_names = {t["name"] for t in doc.document.components["TwoWindingTransformer"]}
    xf_uids = set(br.loc[tr_ratio_rule, "UID"])
    assert xf_names == xf_uids

    hvdc = doc.document.components["TwoTerminalGenericHVDCLine"][0]
    assert hvdc["name"] == "DC1"
    assert hvdc["active_power_limits_from"] == {"min": -100.0, "max": 100.0}
    assert hvdc["active_power_limits_to"] == {"min": -100.0, "max": 100.0}
    assert hvdc["reactive_power_limits_from"] == {"min": 0.0, "max": 100.0}
    assert hvdc["reactive_power_limits_to"] == {"min": 0.0, "max": 100.0}
    assert hvdc["loss"]["value_curve"]["function_data"]["proportional_term"] == 0.1

    doc.validate()


def test_line_a1_spot_check():
    doc = CaseDocument()
    src = data.rts_source_dir() / "SourceData"
    idx = add_topology(doc, src)
    add_branches(doc, src, idx)

    a1 = next(c for c in doc.document.components["Line"] if c["name"] == "A1")
    assert a1["r"] == 0.003
    assert a1["x"] == 0.014
    assert a1["b"]["from"] == 0.2305
    assert a1["b"]["to"] == 0.2305
    assert a1["rating"] == 175.0
    assert a1["rating_b"] == 193.0
    assert a1["rating_c"] == 200.0
    assert a1["base_power"] == 100.0
    assert a1["power_units"] == "NATURAL_UNITS"
    assert a1["parameter_units"] == "COMPONENT_BASE"
    assert a1["angle_limits"] == {"min": -3.1416, "max": 3.1416}

File: julia/RTSGMLCCase/test/test_branches.jl
@testset "branch partition and arcs" begin
    src = joinpath(RTSGMLCCase.rts_source_dir(), RTSGMLCCase.SOURCE_DATA)
    branch = CSV.read(joinpath(src, "branch.csv"), DataFrame)
    dc_branch = CSV.read(joinpath(src, "dc_branch.csv"), DataFrame)
    bus = CSV.read(joinpath(src, "bus.csv"), DataFrame)
    base_kv = Dict(Int(row["Bus ID"]) => Float64(row["BaseKV"]) for row in eachrow(bus))

    doc = RTSGMLCCase.new_document()
    idx = RTSGMLCCase.add_topology!(doc, src)
    RTSGMLCCase.add_branches!(doc, src, idx)

    # Two independent classification rules must agree exactly (R21).
    tr_ratio_rule = [!(Float64(t) in (0.0, 1.0)) for t in branch[!, "Tr Ratio"]]
    voltage_rule = [
        base_kv[Int(row["From Bus"])] != base_kv[Int(row["To Bus"])] for row in eachrow(branch)
    ]
    @test tr_ratio_rule == voltage_rule
    n_xf = count(tr_ratio_rule)

    lines = PowerOpenAPIModels.get_components(doc, "Line")
    circuits = PowerOpenAPIModels.get_components(doc, "TransformerCircuit")
    transformers = PowerOpenAPIModels.get_components(doc, "TwoWindingTransformer")
    hvdc_lines = PowerOpenAPIModels.get_components(doc, "TwoTerminalGenericHVDCLine")
    arcs = PowerOpenAPIModels.get_components(doc, "Arc")

    @test length(lines) == nrow(branch) - n_xf
    @test length(circuits) == n_xf
    @test length(transformers) == n_xf

    # Pinned expected totals (CAMPAIGN-FACTS.md "Expected component counts", R21).
    @test length(lines) == 105
    @test length(circuits) == 15
    @test length(transformers) == 15
    @test length(hvdc_lines) == 1

    # No `TransformerCircuit` carries a `name` -- the name lives on `TwoWindingTransformer`.
    for circuit in circuits
        @test circuit.arc in [a.id for a in arcs]
    end
    for line in lines
        @test line.arc in [a.id for a in arcs]
    end
    xf_names = Set(t.name for t in transformers)
    xf_uids = Set(String(u) for u in branch[tr_ratio_rule, "UID"])
    @test xf_names == xf_uids

    ac_pairs = Set(
        (idx.bus_id_by_number[Int(row["From Bus"])], idx.bus_id_by_number[Int(row["To Bus"])])
        for row in eachrow(branch)
    )
    dc_pairs = Set(
        (idx.bus_id_by_number[Int(row["From Bus"])], idx.bus_id_by_number[Int(row["To Bus"])])
        for row in eachrow(dc_branch)
    )
    @test length(ac_pairs) == 108
    @test length(dc_pairs) == 1
    arc_pairs = [(a.from_id, a.to_id) for a in arcs]
    @test length(Set(arc_pairs)) == length(arc_pairs) # deduplicated
    @test length(union(ac_pairs, dc_pairs)) == length(arcs)
    @test length(arcs) == 109

    # The arc allocator genuinely reuses ids for a known duplicated ordered pair (12 branch.csv
    # rows share their (From Bus, To Bus) with another row) rather than merely landing on the
    # right final count by coincidence.
    a25_1 = only(filter(l -> l.name == "A25-1", lines))
    a25_2 = only(filter(l -> l.name == "A25-2", lines))
    @test a25_1.arc == a25_2.arc

    hvdc = only(hvdc_lines)
    @test hvdc.name == "DC1"
    @test hvdc.active_power_limits_from.min == -100.0
    @test hvdc.active_power_limits_from.max == 100.0
    @test hvdc.active_power_limits_to.min == -100.0
    @test hvdc.active_power_limits_to.max == 100.0
    @test hvdc.reactive_power_limits_from.min == 0.0
    @test hvdc.reactive_power_limits_from.max == 100.0
    @test hvdc.reactive_power_limits_to.min == 0.0
    @test hvdc.reactive_power_limits_to.max == 100.0
    @test hvdc.loss.value_curve.value.function_data.value.proportional_term == 0.1

    PowerOpenAPIModels.validate_document(doc)
end

@testset "line A1 spot check" begin
    src = joinpath(RTSGMLCCase.rts_source_dir(), RTSGMLCCase.SOURCE_DATA)
    doc = RTSGMLCCase.new_document()
    idx = RTSGMLCCase.add_topology!(doc, src)
    RTSGMLCCase.add_branches!(doc, src, idx)

    a1 = only(filter(l -> l.name == "A1", PowerOpenAPIModels.get_components(doc, "Line")))
    @test a1.r == 0.003
    @test a1.x == 0.014
    @test a1.b.from == 0.2305
    @test a1.b.to == 0.2305
    @test a1.rating == 175.0
    @test a1.rating_b == 193.0
    @test a1.rating_c == 200.0
    @test a1.base_power == 100.0
    # R18: power_units/parameter_units are unenforced String aliases in Julia -- assert the
    # emitted literal explicitly.
    @test a1.power_units == PowerOpenAPIModels.UnitSystem("NATURAL_UNITS")
    @test a1.parameter_units == PowerOpenAPIModels.ImpedanceUnitBasis("COMPONENT_BASE")
    @test a1.angle_limits.min == -3.1416
    @test a1.angle_limits.max == 3.1416
end
